### PR TITLE
[Improvement][Spark] Support Local Spark Cluster

### LINKED ISSUE #15548: [Improvement][Spark] Support Local Spark Cluster
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

When a Spark Task executes spark-submit, the 'cluster' and 'client' deploy maps to `--master yarn` or `--master k8s://...`. I would like option to use a local Spark cluster. In other words, the equivalent spark-submit option is:`--master spark://<hostname>:<port>`

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by pegasas**
## Purpose of the pull request

fixes: https://github.com/apache/dolphinscheduler/issues/15548

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Review by songjianet — DISMISSED** [review]

File: dolphinscheduler-task-plugin/dolphinscheduler-task-spark/src/main/java/org/apache/dolphinscheduler/plugin/task/spark/SparkTask.java [hunk 1]
```diff
@@ -124,22 +124,30 @@ protected Map<String, String> getProperties() {
...
         String deployMode = StringUtils.isNotEmpty(sparkParameters.getDeployMode()) ? sparkParameters.getDeployMode()
                 : SparkConstants.DEPLOY_MODE_LOCAL;
 
+        boolean onLocal = SparkConstants.DEPLOY_MODE_LOCAL.equals(deployMode);
         boolean onNativeKubernetes = StringUtils.isNotEmpty(sparkParameters.getNamespace());
 
-        String masterUrl = onNativeKubernetes ? SPARK_ON_K8S_MASTER_PREFIX +
-                Config.fromKubeconfig(taskExecutionContext.getK8sTaskExecutionContext().getConfigYaml()).getMasterUrl()
-                : SparkConstants.SPARK_ON_YARN;
+        String masterUrl = StringUtils.isNotEmpty(sparkParameters.getMaster()) ? sparkParameters.getMaster()
```
Comment (rickchengx) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-spark/src/main/java/org/apache/dolphinscheduler/plugin/task/spark/SparkTask.java:136]: @pegasas  please add UT for this PR.
Also, I think the code here is less readable, WDYT?

Comment (pegasas) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-spark/src/main/java/org/apache/dolphinscheduler/plugin/task/spark/SparkTask.java:136]: > @pegasas please add UT for this PR. Also, I think the code here is less readable, WDYT?

Sure. I will add more UTs for this function.
Maybe we can build a function for this to make it clear?

**Comment by pegasas** [comment]
image
It seems like I have covered my code.
image
image
UT is missed by other code checkin before.

File: dolphinscheduler-task-plugin/dolphinscheduler-task-spark/src/test/java/org/apache/dolphinscheduler/plugin/task/spark/SparkTaskTest.java [hunk 2]
```diff
@@ -41,20 +48,33 @@ public void testBuildCommandWithSparkSql() throws Exception {
...
-                        "--deploy-mode client " +
-                        "--conf spark.driver.cores=1 " +
-                        "--conf spark.driver.memory=512M " +
-                        "--conf spark.executor.instances=2 " +
-                        "--conf spark.executor.cores=2 " +
-                        "--conf spark.executor.memory=1G " +
-                        "--name sparksql " +
-                        "-f /tmp/5536_node.sql",
-                sparkTask.getScript());
+        ResourceContext resourceContext = Mockito.mock(ResourceContext.class);
```
Comment (rickchengx) [inline dolphinscheduler-task-plugin/dolphinscheduler-task-spark/src/test/java/org/apache/dolphinscheduler/plugin/task/spark/SparkTaskTest.java:55]: Overall LGTM, It seems that this ut does not test the case of `spark-submit --master local xxx` or `--master <master-url>`

Comment (pegasas) [reply dolphinscheduler-task-plugin/dolphinscheduler-task-spark/src/test/java/org/apache/dolphinscheduler/plugin/task/spark/SparkTaskTest.java:55]: Sure. I will add more case.

**Comment by pegasas** [comment]
attach test pics:
image
image
image
image

**Comment by rickchengx** [comment]
> attach test pics: image image image image

@pegasas I don't think the `Master` on the ui is a required field, WDYT

**Comment by pegasas** [comment]
> > attach test pics: image image image image
> 
> @pegasas I don't think the `Master` on the ui is a required field, WDYT

Done.

I set required to false first, in the future we may link discussion here.

reference: https://spark.apache.org/docs/latest/submitting-applications.html#master-urls

In previous design, we combines logic with master & deploy-mode.
But in official doc example, it comes out with master urls no matter in local/standalone/yarn/mesos/k8s.

image

**Review by rickchengx — DISMISSED** [review]
LGTM

**Review by fuchanghai — DISMISSED** [review]
LGTM

**Review by ruanwenjun — DISMISSED** [review]
Please update the usage doc.

**Comment by ruanwenjun** [comment]
Can we remove the option 'local' from deploy-mode?

We can only set cluster/client at deploy mode, the local should belong to master config.

**Comment by pegasas** [comment]
> Please update the usage doc.

updated.

**Comment by pegasas** [comment]
> Can we remove the option 'local' from deploy-mode?
> 
> We can only set cluster/client at deploy mode, the local should belong to master config.

Thanks @ruanwenjun .
I‘ve add TODO's in related code lines.

**Comment by rickchengx** [comment]
@pegasas  please Run 'mvn spotless:apply' to fix the code style

**Comment by pegasas** [comment]
> @pegasas please Run 'mvn spotless:apply' to fix the code style

Thanks, Done.
image

**Comment by rickchengx** [comment]
> Can we remove the option 'local' from deploy-mode?
> 
> We can only set cluster/client at deploy mode, the local should belong to master config.

Maybe we should delete `local` option in this PR, WDYT @pegasas

**Comment by pegasas** [comment]
> > Can we remove the option 'local' from deploy-mode?
> > We can only set cluster/client at deploy mode, the local should belong to master config.
> 
> Maybe we should delete `local` option in this PR, WDYT @pegasas

sure.
corrected.

**Review by rickchengx — APPROVED** [review]

**Comment by rickchengx** [comment]
> Please update the usage doc.

PTAL @ruanwenjun

**Pull request merged**